File: Algorithms/Java/quick_sort.java
public class quick_sort {
    public static int partition(int arr[],int low ,int high){
        int pivot = arr[high];
        int i = low-1;
        for(int j = low;j<=high;j++){
            if(arr[j]<pivot){
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        int temp = arr[i+1];
        arr[i+1] = pivot;
        arr[high] = temp;   
        return i+1;
    }
    public static void q_sort(int arr[],int low,int high){
        if(low<high){
            int pi = partition(arr, low, high);
            q_sort(arr, low, pi-1);
            q_sort(arr, pi+1, high);
        }
    }
    public static void main(String[] args){
        int arr[] = {7,45,82,5,72,6,36,8};
        q_sort(arr, 0, arr.length-1);
        for(int i = 0;i<arr.length;i++){
            System.out.println(arr[i]);
        }
    }
}

File: Algorithms/Python/quick_sort.py
from __future__ import annotations

from typing import TypeVar

T = TypeVar("T", bound=int | float)


def partition(arr: list[T], low: int, high: int) -> int:
    """
    Partition the array around the pivot.
    """
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[j], arr[i] = arr[i], arr[j]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quicksort(arr: list[T]) -> list[T]:
    """
    Sort the array using the quicksort algorithm.
    """
    quicksort_helper(arr, 0, len(arr) - 1)
    return arr


def quicksort_helper(arr: list[T], low: int, high: int) -> None:
    """
    Helper function for quicksort.
    """
    if low < high:
        pivot = partition(arr, low, high)
        quicksort_helper(arr, low, pivot - 1)
        quicksort_helper(arr, pivot + 1, high)


if __name__ == "__main__":
    numbers = [1, 5, 2, 7, 3, 9, 4, 6, 8]
    print(quicksort(numbers))  # [1, 2, 3, 4, 5, 6, 7, 8, 9]
